### PR TITLE
Elizabeth/scale refactor

### PR DESCRIPTION
**Pull request by noot**
## Changes
Refactor of scale codec to use an io.Writer for encoding and an io.Reader for decoding.  This also cleans up some of the code (especially in decode, as we no longer need to return bytesDecoded).   Now, instead of returning a byte array for encode and inputting a byte array for decode, we use a io.Writer as the receiver for encoding and an io.Reader as the receiver for decoding.  What this does is allow us to write to an underlying buffer when encoding, then read from a buffer when decoding. 

## Tests:
```
go test ./codec/...
```

### CODE REVIEW

File: codec/decode.go [hunk 1]
```diff
@@ -3,150 +3,170 @@ package codec
...
+}
+
+// decodeSmallInt is used in the DecodeInteger and DecodeBigInteger functions when the mode is <= 2
+// need to pass in the first byte, since we assume it's already been read
+func (sd *Decoder) decodeSmallInt(firstByte byte) (o int64, err error) {
+	mode := firstByte & 3
+	if mode == 0 { // 1 byte mode
+		o = int64(firstByte >> 2)
+	} else if mode == 1 { // 2 byte mode
+		c, err := sd.ReadByte()
```
[the remark this answers is not in the document]
Comment (noot) [reply codec/decode.go]: renamed

File: codec/decode.go [hunk 2]
```diff
@@ -3,150 +3,170 @@ package codec
...
-	byteLen := topSixBits + 4
+	topSixBits := b >> 2
+	byteLen := int(topSixBits) + 4
 
-	if len(b) < int(byteLen)+1 {
-		err = errors.New("could not decode invalid integer")
-	}
-
-	// TODO: use io.Writer to return number of bytesDecoded
+	c := make([]byte, byteLen)
```
Comment (ansermino) [inline codec/decode.go]: still using `c`

Comment (noot) [reply codec/decode.go]: fixed, sorry about force push, accidentally commit trie stuff to this branch so i had to revert

Comment (ansermino) [reply codec/decode.go]: No change

Comment (noot) [reply codec/decode.go]: which line? looks updated?

**Pull request merged**